### ISSUE TITLE
Debian distribution release reported as "NA"

### ISSUE THREAD

**Issue by candlerb** [ansible/ansible#4565]
Using ansible 1.3.3, running `ansible -m setup <host> | grep distribution` against a Debian server.

```
        "ansible_distribution": "Debian",
        "ansible_distribution_release": "NA",
        "ansible_distribution_version": "7.1",
```

I think the distribution_release should report as "wheezy". This information is available if you login to the target system:

```
# lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 7.1 (wheezy)
Release:    7.1
Codename:   wheezy
```

The underlying problem seems to be with the python platform library:

```
>>> import platform
>>> platform.dist()
('debian', '7.1', '')
>>>
>>> platform.linux_distribution()
('debian', '7.1', '')
```

For comparison, if I point ansible -m setup at an Ubuntu 12.04 host, I get:

```
        "ansible_distribution": "Ubuntu",
        "ansible_distribution_release": "precise",
        "ansible_distribution_version": "12.04",

...
>>> import platform
>>> platform.dist()
('Ubuntu', '12.04', 'precise')
>>>
```

**Comment by mpdehaan** [comment]
You seem quite capable with python, would you like to perhaps submit a pull request to fix this?

Just giving the opportunity, otherwise we can take care of it.

**Comment by candlerb** [comment]
I guess there's no point trying to work out why platform.dist() on Debian does't return a release name, and get that fixed upstream, as it isn't going to help manage systems which are already out there.

I did discover after posting this issue that the facts include "ansible_lsb" and the codename is within that:

```
        "ansible_lsb": {
            "codename": "wheezy",
            "description": "Debian GNU/Linux 7.1 (wheezy)",
            "id": "Debian",
            "major_release": "7",
            "release": "7.1"
        },
```

So for my purposes I can just use ansible_lsb.codename.

Perhaps a workaround for ansible would be something along the lines of:

```
try:
    self.facts['distribution_release'] = dist[2] or self.facts['lsb']['codename']
except KeyError:
    self.facts['distribution_release'] = 'NA'
```

This means get_lsb_facts() would have to be called before get_distribution_facts(). To avoid this being a hidden dependency, maybe get_lsb_facts() should be called from within get_distribution_facts() rather than in **init**

**Comment by mpdehaan** [comment]
Yeah, we can't rely on the LSB facts because the lsb tools are not always
installed on all platforms.

The logical thing to do here is make the LSB function return None or
something when there are no facts, and if it got any, to override that
single value.

But we would want to make extra sure it didn't change any other values or
even break casing of existing values.

**Comment by jctanner** [comment]
@mpdehaan are you saying we should change all the 'NA' values in setup to None ?

Current setup module output:

```
        "ansible_distribution_release": "NA", 
        "ansible_product_serial": "NA", 
        "ansible_product_version": "NA", 
```

With None for unfound values:

```
        "ansible_distribution_release": null, 
        "ansible_product_serial": null, 
        "ansible_product_version": null, 
```

**Comment by jctanner** [comment]
TODO:
- do not change all "NA" returns
- return the real release name, not "NA", even if lsb-release is not installed

**Comment by lichesser** [comment]
On Debian 7 (wheezy) and later there is a file `/etc/os-release` which contains the needed data

Should be easy to treat this analogous to other distributions in `get_distribution_facts` in combination with `OSDIST_DICT`.

More info in the Debian bug report http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=659853
And here http://www.freedesktop.org/software/systemd/man/os-release.html

**Comment by lichesser** [comment]
The matching pull request is https://github.com/ansible/ansible/pull/5426

Sorry for the "double" referenced this issue message. The first one happened because I pushed to the wrong branch

**Comment by jctanner** [comment]
Resolved by #5426 and 6e39adb

**Comment by richard-scott** [comment]
I'm getting this on Debian 9.0...
On the first run it is detected ok, and on future runs it is detected as NA ;-(
